### PR TITLE
Amp subscriptions: Read service config from page

### PR DESCRIPTION
**Pull request by prateekbh**
- Reads service config from page
- Waits for service config before initializing
- checks for authorization Url in service config

### CODE REVIEW

**Comment by prateekbh** [comment]
@dvoytenko Looks like renaming changes did not got committed in the last PR, so they are here in this one

File: examples/amp-subscriptions.amp.html [hunk 1]
```diff
@@ -20,6 +20,19 @@
     }
   }
   </script>
+  <script type="json" id="amp-subscriptions">
+    {
+      "services": [
+        {
+          "serviceId":"amp.local.subscription",
```
Comment (dvoytenko) [inline examples/amp-subscriptions.amp.html]: Let's remove `serviceId` here and just default to "local".

File: examples/amp-subscriptions.amp.html [hunk 2]
```diff
@@ -20,6 +20,19 @@
...
+          "authorizationUrl": "/subscription/2/entitlements"
+        },
+        {
+          "serviceId":"google.subscription"
```
Comment (dvoytenko) [inline examples/amp-subscriptions.amp.html]: "subscribe.google.com"

File: extensions/amp-subscriptions/0.1/amp-subscriptions.js [hunk 3]
```diff
@@ -44,11 +47,17 @@ export class SubscriptionService {
...
+    this.serviceConfig_ = null;
+
     /** @private @const {!Array<!SubscriptionPlatform>} */
     this.subscriptionPlatforms_ = [];
 
     /** @private {?EntitlementStore} */
     this.entitlementStore_ = null;
+
+    /** @const @private {!Element} */
+    this.configElement_ = dev().assertElement(configElement);
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/amp-subscriptions.js]: Let's use `user().assertElement` here

File: extensions/amp-subscriptions/0.1/amp-subscriptions.js [hunk 4]
```diff
@@ -58,27 +67,53 @@ export class SubscriptionService {
...
-    ];
-
-    return pageConfigResolver.resolveConfig().then(pageConfig => {
+    const configPromises = Promise.all([
+      this.getServiceConfig_(),
+      pageConfigResolver.resolveConfig(),
+    ]);
+
+    return configPromises.then(promiseValues => {
+      const serviceConfig = user().assert(promiseValues[0]);
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/amp-subscriptions.js]: You have to add typing for `serviceConfig` and `pageConfig`.

File: extensions/amp-subscriptions/0.1/amp-subscriptions.js [hunk 5]
```diff
@@ -58,27 +67,53 @@ export class SubscriptionService {
...
+    });
+  }
+
+  /**
+   * @param {!JsonObject} serviceConfig
+   * @param {!PageConfig} pageConfig
+   * @private
+   */
+  initializeSubscriptionPlatforms_(serviceConfig, pageConfig) {
+    if (serviceConfig['authorizationUrl']) {
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/amp-subscriptions.js]: This should be that `if ((serviceConfig['serviceId'] || 'local') == 'local') {...}`

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js [hunk 6]
```diff
@@ -51,12 +51,13 @@ export class LocalSubscriptionPlatform {
    * @return {!Promise<!Entitlements>}
    */
   getEntitlements() {
-    user().assert(this.serviceConfig_['paywallUrl'],
-        'Service config does not have paywall Url');
+    const authUrl = user().assert(this.serviceConfig_['authorizationUrl'],
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/local-subscription-platform.js]: It's best to initialize all `authorizationUrl` in the constructor and also call `assertHttpsUrl` on it.

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js [hunk 7]
```diff
@@ -51,12 +51,13 @@ export class LocalSubscriptionPlatform {
...
+        'Service config does not have authorization Url');
     const currentProductId = user().assertString(
         this.pageConfig_.getProductId(), 'Current Product ID is null');
+
     return this.xhr_
-        .fetchJson(this.serviceConfig_['paywallUrl'])
+        .fetchJson(authUrl)
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/local-subscription-platform.js]: We need to make sure all CORS stuff is added here.

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js (hunk 6 above)
Comment (prateekbh) [reply extensions/amp-subscriptions/0.1/local-subscription-platform.js]: Done

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js (hunk 7 above)
Comment (prateekbh) [reply extensions/amp-subscriptions/0.1/local-subscription-platform.js]: Is there something like a util for this?

File: extensions/amp-subscriptions/0.1/amp-subscriptions.js (hunk 4 above)
Comment (prateekbh) [reply extensions/amp-subscriptions/0.1/amp-subscriptions.js]: Like a /** typedef */ or a just /** {!PageConfig} *? and /** JsonObject<string, string> */

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js (hunk 7 above)
Comment (dvoytenko) [reply extensions/amp-subscriptions/0.1/local-subscription-platform.js]: It's the `fetchJson` API. I believe it's something like this:
```
fetchJson(url, {
  credentials: 'include',
})
```

But please double-check.

File: extensions/amp-subscriptions/0.1/amp-subscriptions.js (hunk 4 above)
Comment (dvoytenko) [reply extensions/amp-subscriptions/0.1/amp-subscriptions.js]: `/** @type {!PageConfig} */`

Comment (prateekbh) [reply extensions/amp-subscriptions/0.1/amp-subscriptions.js]: Done

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js (hunk 7 above)
Comment (prateekbh) [reply extensions/amp-subscriptions/0.1/local-subscription-platform.js]: Done! I guess!

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js [hunk 8]
```diff
@@ -51,12 +61,14 @@ export class LocalSubscriptionPlatform {
...
    */
   getEntitlements() {
-    user().assert(this.serviceConfig_['paywallUrl'],
-        'Service config does not have paywall Url');
     const currentProductId = user().assertString(
         this.pageConfig_.getProductId(), 'Current Product ID is null');
+
     return this.xhr_
-        .fetchJson(this.serviceConfig_['paywallUrl'])
+        .fetchJson(this.authorizationUrl_,{
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/local-subscription-platform.js]: nit: space between "," and "{"

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js [hunk 9]
```diff
@@ -51,12 +61,14 @@ export class LocalSubscriptionPlatform {
...
+          mode: 'cors',
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/local-subscription-platform.js]: No need for `mode` here. Found this: https://github.com/ampproject/amphtml/blob/master/extensions/amp-access/0.1/amp-access-client.js#L126

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js [hunk 10]
```diff
@@ -51,12 +61,14 @@ export class LocalSubscriptionPlatform {
...
+          credentials: 'include',
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/local-subscription-platform.js]: And this needs a test.

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js (hunk 8 above)
Comment (prateekbh) [reply extensions/amp-subscriptions/0.1/local-subscription-platform.js]: Done

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js (hunk 9 above)
Comment (prateekbh) [reply extensions/amp-subscriptions/0.1/local-subscription-platform.js]: Done

File: extensions/amp-subscriptions/0.1/local-subscription-platform.js (hunk 10 above)
Comment (prateekbh) [reply extensions/amp-subscriptions/0.1/local-subscription-platform.js]: Done

**Pull request merged**